enhancements/olm-curated-catalog.md

```diff
@@ -0,0 +1,302 @@
+---
+title: olm-curated-catalog
+
+authors:
+  - "@cdjohnson"
+
+reviewers:
+
+approvers:
+
+creation-date: 2020-03-19
+
+last-updated: 2020-04-01
+
+status: 
+
+see-also:
+  - "/enhancements/opm-mirroring.md"  
+
+replaces:
+
+superseded-by:
+---
+
+# OLM Curated Catalog
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions
+- Should the filter spec be YAML, JSON, both (auto-detect)?  I think we should support both automatically like kubectl does.
+- Should the filter use patterns/expressions for the channel filters?
+- Verify it's OK that we create new bundle images on the fly.  I don't see a way around this unless unless we remove the bundle image references from the index database.
```

Comment:
When you add a label on your bundle image `operators.operatorframework.io.bundle.channels.v1: stable,alpha`, I read this as saying that this particular bundle belongs to the `stable` and `alpha` channels. So I don't believe that removing a channel (ie. `alpha`) from a particular package requires us to rebuild the bundle.

Reply:
I am somewhat concerned about what happens when you are removing the default package, would you be **required** to specify a default channel? Should this enhancement support configuring the default channel for packages?

Reply:
```
When you add a label on your bundle image operators.operatorframework.io.bundle.channels.v1: stable,alpha, I read this as saying that this particular bundle belongs to the stable and alpha channels. So I don't believe that removing a channel (ie. alpha) from a particular package requires us to rebuild the bundle.
```

This is not strictly true right now. Because some Kube distributions (ex. OpenShift) restrict what OLM can do at runtime with respect to reading image labels, the labels themselves don't drive the package channels. Instead it's driven by the channel data in the `annotations.yaml` file in the `metadata/` folder of the bundle image -- updating that would be required and that would mean an image rebuild.

That being said, there has been some significant discussion around using `opm` to modify an index image. It would mean that the annotations that OLM sees could potentially differ from catalog graph information, but I could see the argument.

Reply:
> I am somewhat concerned about what happens when you are removing the default package, would you be **required** to specify a default channel? Should this enhancement support configuring the default channel for packages?

Yes, this is an important consideration too.  I think we may need to add a "defaultChannel" override to the `filterSpec`.  This isn't really part of the filter per-se, it's really a function of how to build the curated index.

Reply:
wait

> This is not strictly true right now. Because some Kube distributions (ex. OpenShift) restrict what OLM can do at runtime with respect to reading image labels, the labels themselves don't drive the package channels. Instead it's driven by the channel data in the annotations.yaml file in the metadata/ folder of the bundle image -- updating that would be required and that would mean an image rebuild.

why can we not override the bundle intentions for a given index image. I thought that this was in place?

Reply:
There's nothing currently in `opm` that makes this possible. We would need to add controls at the individual bundle level and right now `opm` just takes a list of bundles when adding. It's something we discussed but we have not implemented.

Reply:
> This is not strictly true right now. Because some Kube distributions (ex. OpenShift) restrict what OLM can do at runtime with respect to reading image labels, the labels themselves don't drive the package channels. Instead it's driven by the channel data in the annotations.yaml file in the metadata/ folder of the bundle image -- updating that would be required and that would mean an image rebuild.

My understanding was that what you see in [using the catalog locally](https://github.com/operator-framework/operator-registry#using-the-catalog-locally) is the kind of thing that OLM is doing when it is determining 1) what operators are available and 2) what channels there are.

Specifically:

```
$ grpcurl -plaintext -d '{"name":"etcd"}' localhost:50051 api.Registry/GetPackage
{
  "name": "etcd",
  "channels": [
    {
      "name": "alpha",
      "csvName": "etcdoperator.v0.9.2"
    }
  ],
  "defaultChannelName": "alpha"
}
```

I would be very surprised if OLM was interrogating something other than the registry api for packages, bundles, etc.

I believe what happened is that my original statement wasn't complete and caused confusion.  My position is 1) the database is the source of truth for what operator packages, and their channels, exist in a given catalog 2) removing the channel from the package in the database (or from the index) is necessary and sufficient.

> There's nothing currently in opm that makes this possible. We would need to add controls at the individual bundle level and right now opm just takes a list of bundles when adding. It's something we discussed but we have not implemented.

It feels almost like we are talking about this from different perspectives. As I see it, the bundles have already been created and added to the index. My understanding is that once the bundle is added to the index, the index is the source of truth for all this stuff.

Reply:
Sorry, there's a lot to unpack in this particular train of thought.

Firstly, this is a complex statement:

> It feels almost like we are talking about this from different perspectives. As I see it, the bundles have already been created and added to the index. My understanding is that once the bundle is added to the index, the index is the source of truth for all this stuff.

That is true from the perspective of OLM explicitly for questions about the update graph, but the annotations are not exclusively intended to provide package aggregate information about the bundle, and that information will definitely be used by OLM at runtime. Keep in mind that the Index images built via opm don't store the full bundle manifests, just the head of each default channel to drive the UX -- at runtime OLM actually pulls non default bundles down, unpacks them, and uses those contents to drive install. That is why we still have to parse the contents of the image to handle them. But, we could conceivably come up with a method of separating "graph data" in a way that never needs to be used by OLM, it's just not how it was originally implemented.

Second, it's a question of reproducibility. Indexes are something that are intended to be easily recreated from source (in this case the source being a list of bundle container images or perhaps even the contents of a container image registry. If we modify the graph information ONLY in the index then we are less able to reproduce them.

There has also been some conversation historically about how to back up or recreate an index image from "source", but we haven't explicitly come to a conclusion on the right path forward there (a metadata file? a list of bundle shas? the actual underlying sqlite database?) but each of those has some constraints and tradeoffs that I don't think the project is ready to decide on just yet.

Reply:
> Keep in mind that the Index images built via opm don't store the full bundle manifests, just the head of each default channel to drive the UX -- at runtime OLM actually pulls non default bundles down, unpacks them, and uses those contents to drive install.

need to dive into this a little more. When I go to create a Subscription to an operator (from a catalog backed by an index image), the available and default channels comes from...unpacking all of the relevant bundles? or from the registry api?

> My position is 1) the database is the source of truth for what operator packages, and their channels, exist in a given catalog 2) removing the channel from the package in the database (or from the index) is necessary and sufficient.

This still summarizes where I stand and I have not yet seen sufficient evidence to re-evaluate. If OLM traverses bundles referenced by an index image and a bundle says, "I belong to `unicorn` channel" but there is no record of the `unicorn` channel in the registry...who wins?

Reproducibility is an interesting thing to think about as I didn't know that was the intention of index images but it isn't clear why that means we need to rebuild operator bundle container images.

Reply:
> This still summarizes where I stand and I have not yet seen sufficient evidence to re-evaluate. If OLM traverses bundles referenced by an index image and a bundle says, "I belong to unicorn channel" but there is no record of the unicorn channel in the registry...who wins?

Right now for *this particular bit of metadata* the registry wins, not necessarily for any arbitrary piece of metadata stored in the `/metadata` folder. My point is that right now there does exist this contention and I would like to see that explicitly addressed in this enhancement if we can. I think it's okay to go down that path, but we should state clearly that we are doing it.